### PR TITLE
Update cmake_policy in vcpkg

### PR DESCRIPTION
**Pull request by gwankyun**
Compatibility with versions of CMake older than 3.10 is deprecated.

### CODE REVIEW

File: scripts/buildsystems/vcpkg.cmake [hunk 1]
```diff
@@ -37,7 +37,11 @@ if(CMAKE_VERSION VERSION_LESS Z_VCPKG_CMAKE_REQUIRED_MINIMUM_VERSION)
     message(FATAL_ERROR "vcpkg.cmake requires at least CMake ${Z_VCPKG_CMAKE_REQUIRED_MINIMUM_VERSION}.")
 endif()
 cmake_policy(PUSH)
-cmake_policy(VERSION 3.7.2)
+if(${CMAKE_VERSION} VERSION_LESS 3.31)
+    cmake_policy(VERSION 3.7.2)
+else()
+    cmake_policy(VERSION 3.10)
+endif()
```
Comment (dg0yt) [inline scripts/buildsystems/vcpkg.cmake:44]:
```suggestion
cmake_policy(VERSION 3.7.2...3.10)
```

File: scripts/buildsystems/vcpkg.cmake [hunk 2]
```diff
@@ -875,7 +879,11 @@ macro("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}" z_vcpkg_find_package_package_name)
 endmacro()
 
 cmake_policy(PUSH)
-cmake_policy(VERSION 3.7.2)
+if(${CMAKE_VERSION} VERSION_LESS 3.31)
+    cmake_policy(VERSION 3.7.2)
+else()
+    cmake_policy(VERSION 3.10)
+endif()
```
Comment (dg0yt) [inline scripts/buildsystems/vcpkg.cmake:886]:
```suggestion
cmake_policy(VERSION 3.7.2...3.10)
```

**Comment by BillyONeal** [comment]
Rerunning due to network mistake last night.

**Comment by BillyONeal** [comment]
I think this should be set to 3.16.3
Related: https://github.com/microsoft/vcpkg-docs/pull/415

Tagging vcpkg-team-review to confirm with other maintainers.

**Comment by dg0yt** [comment]
> I think this should be set to 3.16.3

Do you want to raise the general minimum CMake version (as avaiable in Ubuntu 20.04)?
(Just asking.)

**Comment by BillyONeal** [comment]
> Do you want to raise the general minimum CMake version (as avaiable in Ubuntu 20.04)?
> (Just asking.)

That's what I'm arguing, yes. The reason we cared about old CMake versions is "work with the CMake the downstream customer is likely to be using on our supported Linuxes", so with Debian 10 "Buster" leaving support the new oldest distro we care about is Ubuntu 20.04.

File: scripts/buildsystems/vcpkg.cmake [hunk 3]
```diff
@@ -37,7 +37,7 @@ if(CMAKE_VERSION VERSION_LESS Z_VCPKG_CMAKE_REQUIRED_MINIMUM_VERSION)
     message(FATAL_ERROR "vcpkg.cmake requires at least CMake ${Z_VCPKG_CMAKE_REQUIRED_MINIMUM_VERSION}.")
 endif()
 cmake_policy(PUSH)
-cmake_policy(VERSION 3.7.2)
+cmake_policy(VERSION 3.7.2...3.10)
```
Comment (dg0yt) [inline scripts/buildsystems/vcpkg.cmake:40]:
```suggestion
cmake_policy(VERSION 3.16)
```

File: scripts/buildsystems/vcpkg.cmake [hunk 4]
```diff
@@ -875,7 +875,7 @@ macro("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}" z_vcpkg_find_package_package_name)
 endmacro()
 
 cmake_policy(PUSH)
-cmake_policy(VERSION 3.7.2)
+cmake_policy(VERSION 3.7.2...3.10)
```
Comment (dg0yt) [inline scripts/buildsystems/vcpkg.cmake:878]:
```suggestion
cmake_policy(VERSION 3.16)
```

**Review by dg0yt — COMMENTED** [review]
This is what I concluded from the discussion.
BTW it is not necessary to resync with master all the time.

**Comment by BillyONeal** [comment]
@vicroms @ras0219-msft and I discussed this today and affirm that we want this to say 3.16.

**Comment by silverqx** [comment]
Can we merge this? The output is really flooded with these warnings, it's hard to develop with them.

Ubuntu 20.04 LTS has this CMake version.

**Comment by dg0yt** [comment]
There is a change requested but not applied yet.

And there is a set of PRs which all trigger world rebuilds, so it might make sense to coordinate them to reduce impact on users.
- #42045 (ready)
- #41980 (not yet ready)

**Comment by silverqx** [comment]
Ok thats reasonable, but what about cherrypick this change only to main?

**Comment by dg0yt** [comment]
It is not causing *errors*, and there is a mitigation: CMake 3.30.

Again, this PR causes a world rebuild for all users: It invalidates all binary artifacts. That's why it shouldn't occur too often. (And that's the reason why I don't simply open or change another PR. It takes two days to pass CI.) 

Ping @gwankyun

**Comment by patrikhuber** [comment]
Looking forward to getting this merged, I just ran a `sudo apt update/upgrade` on my WSL Ubuntu 24.04, that seemed to have updated CMake to 3.31, and the terminal is now flooded with those warnings, it's hard to find/see anything else.

**Comment by gwankyun** [comment]
> It is not causing _errors_, and there is a mitigation: CMake 3.30.
> 
> Again, this PR causes a world rebuild for all users: It invalidates all binary artifacts. That's why it shouldn't occur too often. (And that's the reason why I don't simply open or change another PR. It takes two days to pass CI.)
> 
> Ping @gwankyun

OK, I'm not familiar with the PR process.

**Review by dg0yt — COMMENTED** [review]
LGTM if the only failing port is cmake-user.

cmake-user is a test port, located in `scripts/test_ports`. It tests standard `find_package` for the current vcpkg cmake (3.30) and for the minimum supported version. That's why it fails: It still tests with 3.7 binaries, but the toolchain wants 3.16 now. 

So this port needs to be updated, too. Either you do it in this PR, or I do it separately. But I don't know if I will have the time to do so before sunday.

**Comment by dg0yt** [comment]
For a minimal fix, you only have to remove cmake 3.7 testing from cmake-user's default features, i.e. these lines:
https://github.com/microsoft/vcpkg/blob/9b5cb8e554487f3edb5d50b80188883846f81e14/scripts/test_ports/cmake-user/vcpkg.json#L21-L27

This keeps testing with cmake 3.30 while disabling the outdated part.
I can complete remove `cmake-3-7` and add `cmake-3-16` in a separate PR.

**Comment by gwankyun** [comment]
@microsoft-github-policy-service agree

**Comment by Osyotr** [comment]
There are a few other things (mostly comments) that can be removed after rasing minimum CMake version to 3.16.
https://github.com/microsoft/vcpkg/blob/b2cb0da531c2f1f740045bfe7c4dac59f0b2b69c/scripts/buildsystems/vcpkg.cmake#L4-L5

Not sure it's worth restting CI over this.

**Review by BillyONeal — DISMISSED** [review]
I fixed the 3-7 test rather than deactivating it.

**Comment by BillyONeal** [comment]
I'm reverting the cmake-user changes because we merged https://github.com/microsoft/vcpkg/pull/42203

**Pull request merged**